File: lint-arwaky/archive/surfaces/mcp_job_handler.py
import json
from typing import Any
from ..contract import JobRegistryAggregate, ServiceContainerAggregate


class McpJobCommandsSurface:
    """Surface for MCP job management tools."""

    def __init__(self, mcp: Any = None) -> None:
        self.mcp = mcp
        self.container: ServiceContainerAggregate | None = None

    def register_all(self, container: ServiceContainerAggregate) -> None:
        """Register the job management tools."""
        self.container = container
        if self.mcp:
            self.mcp.tool()(self.check_status)
            self.mcp.tool()(self.cancel_job)

    async def check_status(self, job_id: str | None = None):
        """Check status of long-running lint jobs."""
        if self.container is None:
            raise RuntimeError("Container not initialized")
        job_registry = self.container.get(JobRegistryAggregate)

        if job_id is None:
            all_jobs_vo = await job_registry.list_jobs()
            # Handle both raw dict and MetadataVO
            all_jobs = (
                all_jobs_vo.value if hasattr(all_jobs_vo, "value") else all_jobs_vo
            )
            jobs_list = [
                {"job_id": str(jid), "status": info["status"], "action": info["action"]}
                for jid, info in all_jobs.items()
            ]
            return json.dumps({"jobs": jobs_list, "total": len(jobs_list)})

        job_info_vo = await job_registry.get_job(job_id)
        if job_info_vo is None:
            return json.dumps(
                {"error": f"Job '{job_id}' not found", "status": "not_found"}
            )

        job_info = job_info_vo.value if hasattr(job_info_vo, "value") else job_info_vo
        return json.dumps(
            {
                "job_id": job_id,
                "status": job_info["status"],
                "action": job_info["action"],
                "started_at": job_info.get("started_at"),
                "completed_at": job_info.get("completed_at"),
                "result": job_info.get("result"),
            }
        )

    async def cancel_job(self, job_id: str):
        """Cancel a running lint job."""
        if self.container is None:
            raise RuntimeError("Container not initialized")
        job_registry = self.container.get(JobRegistryAggregate)

        from ..taxonomy import JobId

        success = await job_registry.cancel_job(JobId(value=job_id))

        return json.dumps(
            {
                "job_id": job_id,
                "status": "cancelled" if success else "failed_to_cancel",
                "success": bool(success),
            }
        )


def register_job_commands(mcp, container: ServiceContainerAggregate) -> None:
    """Factory function for container-aware registration."""
    surface = McpJobCommandsSurface(mcp)
    surface.register_all(container)


def register_check_status(mcp, container: ServiceContainerAggregate) -> None:
    """Legacy wrapper updated to be container-aware."""
    register_job_commands(mcp, container)


def register_cancel_job(mcp, container: ServiceContainerAggregate) -> None:
    """Legacy wrapper updated to be container-aware."""
    register_job_commands(mcp, container)

File: lint-arwaky/src-rust/surfaces/mcp_job_handler.rs
use crate::contract::job_registry_aggregate::JobRegistryAggregate;
use crate::taxonomy::job_action_vo::JobId;
use serde_json::json;
use std::sync::Arc;

pub struct McpJobCommandsSurface;

impl McpJobCommandsSurface {
    pub fn new() -> Self {
        Self
    }

    pub async fn check_status(
        &self,
        container: &Arc<crate::contract::service_container_aggregate::ServiceContainerAggregate>,
        job_id: Option<String>,
    ) -> Result<String, String> {
        let job_registry = container
            .get::<Arc<dyn JobRegistryAggregate>>()
            .ok_or_else(|| "Container not initialized".to_string())?;

        match job_id {
            None => {
                let all_jobs_vo = job_registry.list_jobs().await.map_err(|e| e.to_string())?;
                let all_jobs = all_jobs_vo;
                let jobs_list: Vec<serde_json::Value> = all_jobs
                    .iter()
                    .map(|(jid, info)| {
                        json!({
                            "job_id": jid,
                            "status": info.status,
                            "action": info.action,
                        })
                    })
                    .collect();
                Ok(json!({ "jobs": jobs_list, "total": jobs_list.len() }).to_string())
            }
            Some(jid) => {
                let job_info = job_registry
                    .get_job(&JobId::new(&jid))
                    .await
                    .map_err(|e| e.to_string())?;
                match job_info {
                    Some(info) => Ok(json!({
                        "job_id": jid,
                        "status": info.status,
                        "action": info.action,
                        "started_at": info.started_at,
                        "completed_at": info.completed_at,
                        "result": info.result,
                    })
                    .to_string()),
                    None => Ok(json!({
                        "error": format!("Job '{}' not found", jid),
                        "status": "not_found",
                    })
                    .to_string()),
                }
            }
        }
    }

    pub async fn cancel_job(
        &self,
        container: &Arc<crate::contract::service_container_aggregate::ServiceContainerAggregate>,
        job_id: String,
    ) -> Result<String, String> {
        let job_registry = container
            .get::<Arc<dyn JobRegistryAggregate>>()
            .ok_or_else(|| "Container not initialized".to_string())?;

        let success = job_registry
            .cancel_job(&JobId::new(&job_id))
            .await
            .map_err(|e| e.to_string())?;

        Ok(json!({
            "job_id": job_id,
            "status": if success { "cancelled" } else { "failed_to_cancel" },
            "success": success,
        })
        .to_string())
    }
}
